vault/expiration.go

```diff
@@ -505,6 +505,28 @@ func (m *ExpirationManager) revokeCommon(leaseID string, force, skipToken bool)
 		return nil
 	}
 
+	// If not in force mode, where we want to ignore errors, and not in
+	// skiptoken mode where we're already in the middle of a revocation
+	// operation, if the expiration time has not already been hit, set
+	// expiration to now, persist, and call updatePending to hand off
+	// revocation to the expiration manager's pending timer map. We still
+	// attempt to continue revocation here.
+	if !force && !skipToken && le.ExpireTime.After(time.Now()) {
```

Comment:
I am not sure if this might end up being the same, but use pendingLock's lock to set the ExpireTime which should trigger on the the original revokeCommon goroutine (and not the one called by updatePendingInternal), thus avoiding the race.
```
if !force && !skipToken {
    m.pendingLock.Lock()
    if le.ExpireTime.Before(time.Now()) {
        return nil
    }
    le.ExpireTime = time.Now()
    ...
}
```

Reply:
AfterFunc always runs in its own goroutine, see https://golang.org/pkg/time/#AfterFunc

Reply:
Yea I was talking about race between the original and the one created by AfterFunc (sort of the same comment left by Brian in slack).

Reply:
As Jeff pointed out the `return nil` below ensures there isn't a race

Reply:
Yea I was wondering if we needed to do the time check with a lock held, but since we update and persist `le.ExpireTime` before calling `updatePendingInternal` it shouldn't matter.